core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/filter_store.go

```diff
@@ -0,0 +1,171 @@
+package logprovider
+
+import (
+	"math/big"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"golang.org/x/time/rate"
+)
+
+type UpkeepFilterStore interface {
+	InitializeActiveUpkeeps(filters ...upkeepFilter)
+	GetIDs(selector func(upkeepFilter) bool) []*big.Int
+	UpdateFilters(updater func(upkeepFilter, upkeepFilter) upkeepFilter, filters ...upkeepFilter)
+	Has(id *big.Int) bool
+	RangeFiltersByIDs(iterator func(int, upkeepFilter), ids ...*big.Int)
+	GetFilters(selector func(upkeepFilter) bool) []upkeepFilter
+	AddActiveUpkeeps(filters ...upkeepFilter)
+	RemoveActiveUpkeeps(filters ...upkeepFilter)
+	Size() int
+}
+
+var _ UpkeepFilterStore = &upkeepFilterStore{}
+
+type upkeepFilter struct {
+	addr     []byte
+	topics   []common.Hash
+	upkeepID *big.Int
+	// lastPollBlock is the last block number the logs were fetched for this upkeep
+	// used by log event provider.
+	lastPollBlock int64
+	// blockLimiter is used to limit the number of blocks to fetch logs for an upkeep.
+	// used by log event provider.
+	blockLimiter *rate.Limiter
+	// lastRePollBlock is the last block number the logs were recovered for this upkeep
+	// used by log recoverer.
+	lastRePollBlock int64
+}
+
+type upkeepFilterStore struct {
+	lock    *sync.RWMutex
+	filters map[string]upkeepFilter
+}
+
+func NewUpkeepFilterStore() *upkeepFilterStore {
+	return &upkeepFilterStore{
+		lock:    &sync.RWMutex{},
+		filters: map[string]upkeepFilter{},
+	}
+}
+
+func (s *upkeepFilterStore) InitializeActiveUpkeeps(filters ...upkeepFilter) {
+	s.lock.Lock()
+	s.filters = make(map[string]upkeepFilter)
+	s.lock.Unlock()
+
+	s.AddActiveUpkeeps(filters...)
+}
+
+func (s *upkeepFilterStore) GetIDs(selector func(upkeepFilter) bool) []*big.Int {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	if selector == nil {
+		// noop selector returns true for all filters
+		selector = func(upkeepFilter) bool { return true }
+	}
+
+	var ids []*big.Int
+	for _, f := range s.filters {
+		if selector(f) {
+			ids = append(ids, big.NewInt(0).Add(big.NewInt(0), f.upkeepID))
```

Comment:
nit I might suggest adding something like a `clone()` function to  `core/utils/big.go` and using that here - this confused me a bit at first